### ISSUE TITLE
SB 404 on some urls

### ISSUE THREAD

**Issue by Edocsil** [JimmXinu/FanFicFare#438]
After the update I've seen that SB updates always give an error and proceed to do a full download (which it does just fine). I have also found out that whenever the IMAP update can't find a file and outputs a "continuing with update= false" it download in full not just that one but every story after that in the batch.

```
[21:15] Edoc@Termux:~ $ dff https://forums.spacebattles.com/posts/62067282
FFF: DEBUG: 2019-10-20 21:15:51,631: cli.py(184):     OS Version:Linux-4.4.153-perf+-aarch64-with-libc
FFF: DEBUG: 2019-10-20 21:15:51,632: cli.py(185): Python Version:3.7.4 (default, Jul 28 2019, 22:33:35)
[Clang 8.0.7 (https://android.googlesource.com/toolchain/clang b55f2d4ebfd35bf6
FFF: DEBUG: 2019-10-20 21:15:51,632: cli.py(186):    FFF Version:3.12.0
FFF: INFO: 2019-10-20 21:15:51,796: base_xenforoforum_adapter.py(420): url: https://forums.spacebattles.com/pos
ts/62067282/
FFF: DEBUG: 2019-10-20 21:15:51,798: configurable.py(1133): #####################################
pagecache(GET) MISS: https://forums.spacebattles.com/posts/62067282/
FFF: DEBUG: 2019-10-20 21:16:01,654: configurable.py(1001): Encoding:utf8
FFF: INFO: 2019-10-20 21:16:01,661: base_xenforoforum_adapter.py(425): use useurl: https://forums.spacebattles.
com/threads/mothercrystal-worm-jrpg-crossover.759331/page-18#post-62067282
FFF: DEBUG: 2019-10-20 21:16:01,676: configurable.py(1133): #####################################
pagecache(GET) MISS: https://forums.spacebattles.com/login
FFF: DEBUG: 2019-10-20 21:16:06,418: configurable.py(1001): Encoding:utf8
FFF: DEBUG: 2019-10-20 21:16:06,424: base_xenforo2forum_adapter.py(79): Will now login to URL (https://forums.s
pacebattles.com/login/login) as (*****)
FFF: DEBUG: 2019-10-20 21:16:06,429: configurable.py(1062): #####################################
pagecache(POST) MISS: https://forums.spacebattles.com/login/login?_xfRedirect=https://forums.spacebattles.com/&
_xfToken=1571598967,d03c781a0e6b9f67510d3ebb705ce281&login=XXXXXXXX&password=XXXXXXXX&remember=1?
FFF: DEBUG: 2019-10-20 21:16:12,044: configurable.py(1001): Encoding:utf8
FFF: DEBUG: 2019-10-20 21:16:12,063: configurable.py(1133): #####################################
pagecache(GET) MISS: https://forums.spacebattles.com/threads/mothercrystal-worm-jrpg-crossover.759331/page-18%2
3post-62067282
FFF: DEBUG: 2019-10-20 21:16:18,783: configurable.py(1213): Caught an exception reading URL: https://forums.spa
cebattles.com/threads/mothercrystal-worm-jrpg-crossover.759331/page-18#post-62067282  Exception HTTP Error 404:
 Not Found.
FFF: DEBUG: 2019-10-20 21:16:18,786: configurable.py(1220): Giving up on https://forums.spacebattles.com/thread
s/mothercrystal-worm-jrpg-crossover.759331/page-18#post-62067282
FFF: DEBUG: 2019-10-20 21:16:18,787: configurable.py(1221): HTTP Error 404: Not Found
NoneType: None
FFF: DEBUG: 2019-10-20 21:16:18,798: configurable.py(1133): #####################################
pagecache(GET) MISS: https://forums.spacebattles.com/login
FFF: DEBUG: 2019-10-20 21:16:23,212: configurable.py(1001): Encoding:utf8
FFF: DEBUG: 2019-10-20 21:16:23,217: base_xenforo2forum_adapter.py(79): Will now login to URL (https://forums.s
pacebattles.com/login/login) as (*****)
FFF: DEBUG: 2019-10-20 21:16:23,223: configurable.py(1062): #####################################
pagecache(POST) MISS: https://forums.spacebattles.com/login/login?_xfRedirect=https://forums.spacebattles.com/&
_xfToken=1571598983,f807193e818de76414d376beb546388e&login=XXXXXXXX&password=XXXXXXXX&remember=1?Accept=text/ht
ml,*/*&Content-type=application/x-www-form-urlencoded
FFF: DEBUG: 2019-10-20 21:16:28,021: configurable.py(1001): Encoding:utf8
FFF: DEBUG: 2019-10-20 21:16:28,034: configurable.py(1133): #####################################
pagecache(GET) MISS: https://forums.spacebattles.com/threads/mothercrystal-worm-jrpg-crossover.759331/page-18%2
3post-62067282
FFF: DEBUG: 2019-10-20 21:16:34,477: configurable.py(1213): Caught an exception reading URL: https://forums.spa
cebattles.com/threads/mothercrystal-worm-jrpg-crossover.759331/page-18#post-62067282  Exception HTTP Error 404:
 Not Found.
FFF: DEBUG: 2019-10-20 21:16:34,478: configurable.py(1220): Giving up on https://forums.spacebattles.com/thread
s/mothercrystal-worm-jrpg-crossover.759331/page-18#post-62067282
FFF: DEBUG: 2019-10-20 21:16:34,478: configurable.py(1221): HTTP Error 404: Not Found
Traceback (most recent call last):
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/adapters/base_xenforoforum_adapt
er.py", line 429, in extractChapterUrlsAndMetadata
    (data,opened) = self._fetchUrlOpened(useurl)
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/adapters/base_xenforoforum_adapt
er.py", line 103, in _fetchUrlOpened
    referer=referer)
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/configurable.py", line 1321, in 
_fetchUrlOpened
    referer=referer)
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/configurable.py", line 1222, in 
_fetchUrlOpened
    raise(excpt)
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/configurable.py", line 1208, in 
_fetchUrlOpened
    referer=referer)
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/configurable.py", line 1167, in 
_fetchUrlRawOpened
    float(self.getConfig('connect_timeout',30.0)))
  File "/data/data/com.termux/files/usr/lib/python3.7/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/data/data/com.termux/files/usr/lib/python3.7/urllib/request.py", line 641, in http_response
    'http', request, response, code, msg, hdrs)
  File "/data/data/com.termux/files/usr/lib/python3.7/urllib/request.py", line 569, in error
    return self._call_chain(*args)
  File "/data/data/com.termux/files/usr/lib/python3.7/urllib/request.py", line 503, in _call_chain
    result = func(*args)
  File "/data/data/com.termux/files/usr/lib/python3.7/urllib/request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 404: Not Found
Traceback (most recent call last):
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/adapters/base_xenforoforum_adapt
er.py", line 429, in extractChapterUrlsAndMetadata
    (data,opened) = self._fetchUrlOpened(useurl)
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/adapters/base_xenforoforum_adapt
er.py", line 103, in _fetchUrlOpened
    referer=referer)
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/configurable.py", line 1321, in 
_fetchUrlOpened
    referer=referer)
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/configurable.py", line 1222, in 
_fetchUrlOpened
    raise(excpt)
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/configurable.py", line 1208, in 
_fetchUrlOpened
    referer=referer)
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/configurable.py", line 1167, in 
_fetchUrlRawOpened
    float(self.getConfig('connect_timeout',30.0)))
  File "/data/data/com.termux/files/usr/lib/python3.7/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/data/data/com.termux/files/usr/lib/python3.7/urllib/request.py", line 641, in http_response
    'http', request, response, code, msg, hdrs)
  File "/data/data/com.termux/files/usr/lib/python3.7/urllib/request.py", line 569, in error
    return self._call_chain(*args)
  File "/data/data/com.termux/files/usr/lib/python3.7/urllib/request.py", line 503, in _call_chain
    result = func(*args)
  File "/data/data/com.termux/files/usr/lib/python3.7/urllib/request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 404: Not Found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/data/com.termux/files/usr/bin/fanficfare", line 11, in <module>
    load_entry_point('FanFicFare==3.12.0', 'console_scripts', 'fanficfare')()
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/cli.py", line 290, in main
    passed_personalini)
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/cli.py", line 402, in do_downloa
d
    adapter.getStoryMetadataOnly()
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/adapters/base_adapter.py", line 
296, in getStoryMetadataOnly
    self.doExtractChapterUrlsAndMetadata(get_cover=get_cover)
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/adapters/base_adapter.py", line 
380, in doExtractChapterUrlsAndMetadata
    return self.extractChapterUrlsAndMetadata()
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/adapters/base_xenforoforum_adapt
er.py", line 436, in extractChapterUrlsAndMetadata
    (data,opened) = self._fetchUrlOpened(useurl)
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/adapters/base_xenforoforum_adapt
er.py", line 103, in _fetchUrlOpened
    referer=referer)
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/configurable.py", line 1321, in 
_fetchUrlOpened
    referer=referer)
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/configurable.py", line 1222, in 
_fetchUrlOpened
    raise(excpt)
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/configurable.py", line 1208, in 
_fetchUrlOpened
    referer=referer)
  File "/data/data/com.termux/files/usr/lib/python3.7/site-packages/fanficfare/configurable.py", line 1167, in 
_fetchUrlRawOpened
    float(self.getConfig('connect_timeout',30.0)))
  File "/data/data/com.termux/files/usr/lib/python3.7/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/data/data/com.termux/files/usr/lib/python3.7/urllib/request.py", line 641, in http_response
    'http', request, response, code, msg, hdrs)
  File "/data/data/com.termux/files/usr/lib/python3.7/urllib/request.py", line 569, in error
    return self._call_chain(*args)
  File "/data/data/com.termux/files/usr/lib/python3.7/urllib/request.py", line 503, in _call_chain
    result = func(*args)
  File "/data/data/com.termux/files/usr/lib/python3.7/urllib/request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 404: Not Found
[21:16] Edoc@Termux:~ $
```

**Comment by JimmXinu** [comment]
You actually have three unrelated issues here, I think.

> After the update I've seen that SB updates always give an error and proceed to do a full download (which it does just fine).

SB changed to xenforo 2 recently, just before the FFF release, in fact.  I haven't seen anything like that, but I confess I use the Calibre version exclusively for my own use.  I'd need to see some debug output and exact steps to know what's going on there.  Open a separate issue if you want to pursue it further.

> I have also found out that whenever the IMAP update can't find a file and outputs a "continuing with update= false" it download in full not just that one but every story after that in the batch.

On a failure to update, the CLI flips the 'update' option off entirely, which I confess is not ideal for batch processing.  I can address that(939a8b050a93717cba4e2392faa0df6f22df08e7), but not *why* the update failed, since you didn't include the exception.  I assume it's related to either the 404 below or the unknown issue above.

Re: The debug output which is appears unrelated to the above issues:  `dff` is not a executable I provide.  I assume it is an alias for `fanficfare -d`.

This 404 issue only expresses with `always_login:true` or `fetch_last_page:true`.  `fetch_last_page` is being removed from the next version, FYI.  a96c5aa80071a70bd5a828b69a6529d7f0ec7aae

I've uploaded a test version of the CLI with these changes:
`pip install --extra-index-url https://testpypi.python.org/pypi --upgrade FanFicFare`

**Comment by Edocsil** [comment]
Sorry if I wasn't that clear, it's only two issues. The SB issue was the one in the debug output (404), when that happened in a batch it would just do a full download which it had no problem doing, I just pasted the first URL that failed at that moment and did a debug download in order to have something to show (I'll try to run the IMAP download in debug the next few days, I always forget).
Together with the other issue this made almost every single batch download take very long so I had to do them by hand.

I have already updated to 3.12.3 and will report in the next few days if updates fail again. Thanks a lot!
